### PR TITLE
Feature/gsl develop diagnostics 20210609

### PR DESCRIPTION
**Pull request by SamuelTrahanNOAA**
# PR Checklist

- [x] Ths PR is up-to-date with the top of all sub-component repositories except for those sub-components which are the subject of this PR. Please consult the ufs-weather-model [wiki](https://github.com/ufs-community/ufs-weather-model/wiki/Making-code-changes-in-the-UFS-weather-model-and-its-subcomponents) if you are unsure how to do this.

- [x] This PR has been tested using a branch which is up-to-date with the top of all sub-component repositories except for those sub-components which are the subject of this PR

- [x] An Issue describing the work contained in this PR has been created either in the subcomponent(s) or in the ufs-weather-model. The Issue should be created in the repository that is most relevant to the changes in contained in the PR. The Issue and the dependent sub-component PR 
are specified below.

- [x] **N/A** If new or updated input data is required by this PR, it is clearly stated in the text of the PR.

## Description

There are three changes:

1. Change the timestep for the fv3_gsd and fv3_gsd_debug tests. Parallels use this new timestep to avoid crashes.
2. Per-timestep surface pressure diagnostics, similar to WRF
3. Diagnostic 3D tendencies for all tracers.

Diagnostic 3D tendencies are documented in detail in this ccpp-doc PR:

https://github.com/NCAR/ccpp-doc/pull/38/files

The per-timestep surface pressure diagnostics are controlled by a new namelist parameter, `print_diff_pgr`, which is off by default. The output looks similar to the WRF-ARW:

```
  0: At forecast hour     0.600 mean abs pgr change is       0.86422375 hPa/hr
  0:   max abs change      0.0016595177 bar  at  tile=4 i=9 j=26
  0:   pgr at that point   1.0098246074 bar      lat=   35.881699 lon=  147.231452
```

### Issue(s) addressed

gsl physics regtest timestep: https://github.com/ufs-community/ufs-weather-model/issues/661
3D diagnostic tendencies: https://github.com/ufs-community/ufs-weather-model/issues/662
per-timestep diagnostics: https://github.com/ufs-community/ufs-weather-model/issues/663

## Testing

> How were these changes tested? What compilers / HPCs was it tested with? Are the changes covered by regression tests? (If not, why? Do new tests need to be added?) Have regression tests and unit tests (utests) been run? On which platforms and with which compilers? (Note that unit tests can only be run on tier-1 platforms)

Eight CCPP suites were tested with all diagnostics enabled. This led to 30-90 variables, all of which were examined visually. A more quantitative comparison, comparing the accumulated change to the total change, proved to be difficult since 64-bit floating point is not sufficient for many processes. (That was confirmed experimentally.) However, these tests matched as much as could be expected, given the limitations of precision. Also, the full suite of regression tests were run to ensure non-diagnostic fields won't change.

There are two new tests: control_diag_debug and fv3_gsd_debug. They differ from their non-diag version only in the input.nml, diag_table, forecast length, and output. The fv3_gsd and fv3_gsd_debug tests' output has changed because of the new timestep. All other regression test output should remain the same.

- [x] hera.intel
- [x] hera.gnu
- [x] orion.intel
- [x] cheyenne.intel 
- [x] cheyenne.gnu
- [x] gaea.intel 
- [x] jet.intel
- [ ] wcoss_cray: Skip due to unavailability
- [ ] wcoss_dell_p3: Skip due to unavailability
- [x] CI: https://github.com/ufs-community/ufs-weather-model/actions/runs/1049611049

## Dependencies

> If testing this branch requires non-default branches in other repositories, list them. Those branches should have matching names (ideally).

> Do PRs in upstream repositories need to be merged first?
> If so add the "waiting for other repos" label and list the upstream PRs

https://github.com/NCAR/ccpp-physics/pull/681
https://github.com/NOAA-EMC/fv3atm/pull/332

### CODE REVIEW

**Comment by junwang-noaa** [comment]
Sam, can you create an issue on ufs-weather-model and list the issue number
in the "issue addressed" section? Please see the PR checklist item 3 above.

**Comment by SamuelTrahanNOAA** [comment]
> Sam, can you create an issue on ufs-weather-model and list the issue number in the "issue addressed" section.

Done.

File: tests/parm/control_ca.nml.IN
Comment (junwang-noaa) [inline tests/parm/control_ca.nml.IN]: May I ask why those lines are added? Those values are the default values in the model code.

Comment (SamuelTrahanNOAA) [reply tests/parm/control_ca.nml.IN]: I don't know why that file was changed; some of these changes did not come from me. I'll revert that file and see if the regression tests pass.

Comment (SamuelTrahanNOAA) [reply tests/parm/control_ca.nml.IN]: Those four lines have been removed.

The later lines, to add `ldiag3d=@[LDIAG3D]`, etc. are intentional, and will be set to the model defaults if diagnostics are disabled, or turned on if diagnostics are enabled.

**Comment by BrianCurtis-NOAA** [comment]
Machine: gaea
Compiler: intel
Job: BL
Repo location: /lustre/f2/pdata/ncep/emc.nemspara/autort/pr/676578304/20210720134506/ufs-weather-model
Please make changes and add the following label back:
gaea-intel-BL

**Comment by BrianCurtis-NOAA** [comment]
Machine: hera
Compiler: gnu
Job: BL
Repo location: /scratch1/NCEPDEV/nems/emc.nemspara/autort/pr/676578304/20210720164508/ufs-weather-model
Please manually delete: /scratch1/NCEPDEV/stmp2/emc.nemspara/FV3_RT/rt_115096
Baseline creation and move successful
Repo location: /scratch1/NCEPDEV/nems/emc.nemspara/autort/pr/676578304/20210720172424/ufs-weather-model
Please manually delete: /scratch1/NCEPDEV/stmp2/emc.nemspara/FV3_RT/rt_139929
Cannot upload hera.gnu RT LogIt is blocked by PR owner
Please obtain logs from /scratch1/NCEPDEV/nems/emc.nemspara/autort/pr/676578304/20210720172424/ufs-weather-model

**Comment by MinsukJi-NOAA** [comment]
@climbfuji verification runs are in progress on jet right now, so I removed your label

**Review by climbfuji — APPROVED** [review]
I had reviewed these changes in the past when they were merged into gsl/develop, still looking good.

**Comment by BrianCurtis-NOAA** [comment]
Cheyenne GNU RT log is on HERA at: /scratch1/NCEPDEV/nems/Brian.Curtis

I'll have the INTEL one there too once it's done. I'll comment here when it is.

File: tests/tests/fv3_gsd_drag_suite
Comment (junwang-noaa) [inline tests/tests/fv3_gsd_drag_suite:79]: I am curious what is the reason to reduce the time step in this PR? I seem not seeing any code changes that require this change.

Comment (SamuelTrahanNOAA) [reply tests/tests/fv3_gsd_drag_suite:79]: This is unrelated to the diagnostics. The gsd physics suite parallels needed a smaller timestep to avoid crashes, and we're updating the regression tests to match.

Comment (climbfuji) [reply tests/tests/fv3_gsd_drag_suite:79]: Yes.

**Review by junwang-noaa — APPROVED** [review]
Need to point back to fv3 develop branch after fv3 PR is merged.

**Comment by BrianCurtis-NOAA** [comment]
Cheyenne INTEL logs are completed and moved to Hera for you.

**Comment by climbfuji** [comment]
Ready to merge ccpp-physics? Please let me know. Thanks!

**Comment by SamuelTrahanNOAA** [comment]
> Ready to merge ccpp-physics? Please let me know. Thanks!

Who are you asking? I'm certainly ready. I don't see any `stampede.intel` test logs though.

**Comment by climbfuji** [comment]
> > Ready to merge ccpp-physics? Please let me know. Thanks!
> 
> Who are you asking? I'm certainly ready. I don't see any `stampede.intel` test logs though.

Jun / Minsuk / Dusan.

Stampede is not a tier-1 platform, no tests are run there.

**Comment by MinsukJi-NOAA** [comment]
> Stampede is not a tier-1 platform, no tests are run there.

Yes, that is correct.

**Comment by SamuelTrahanNOAA** [comment]
I updated the branch to point to the NOAA-EMC fv3atm, now that its PR is merged.

**Review by DusanJovic-NOAA — APPROVED** [review]

**Comment by MinsukJi-NOAA** [comment]
> I updated the branch to point to the NOAA-EMC fv3atm, now that its PR is merged.

Looks good

**Pull request merged**
